### DISCUSSION TITLE
Irrigazione: ce l'ho fatta!

### DISCUSSION THREAD

**Discussion by riccionese** [indomus/forum#584] [GENERALE - Discussioni generiche]
Alla fine, un po' rubando idee da altri e un po' studiando e sperimentando nei ritagli di tempo, penso di avercela fatta.
Con un Sonoff CH4ProR3 tasmotizzato, che pilota 4 valvole di irrigazione a 24v (ho preso un alimentatore specifico), un Sonoff mini, anche lui con tasmota, che ho collegato al rain-click per non far partire i programmi in caso sia piovuto, sono riuscito finalmente a scrivere un package che:
- ha due programmi distinti.
- per ogni programma posso farli partire indistintamente
- - ad un ora specifica
- - in funzione dell'alba (per ora impostata fisso a 30 minuti prima dell'alba)
- - ogni X giorni all'ora specifica di cui sopra
- - oppure manualmente 
- inoltre per ogni programma posso ovviamente definire i minuti per ogni valvola ed i giorni in cui fare o non fare partire l'irrigazione

ovviamente ho anche il "ferma tutto" manuale ed i controlli diretti sulle singole valvole ed il reset delle valvole nel caso HA si riavvii.

Sull'estetica della plancia credo di doversi ancora lavorare ma quantomeno mi sembra "pratica"

![Irrigazione](https://user-images.githubusercontent.com/62329292/127778395-b35b8c29-2147-4d8d-8089-be8fa4974edb.jpg)

Opinioni o suggerimenti sono graditi
ciao

**Comment by Maspar79** [comment]
Ciao Riccionese, parlavi di package ma non ho capito da dove dovremmo scaricarlo per poterlo provare, altrimenti che suggerimenti dovremmo darti?

**Reply by riccionese** [reply]
Ciao, 
scusa ma siccome avevo rotto un po' le scatole qui sul "Discussion" con varie domande, ho pensato di comunicare che alla fine ho raggiunto il risultato "funzionale" che mi serviva.
Ieri ho tolto tutti i dispositivi "dal banco" e li ho montati in operativo: l'irrigazione funziona correttamente.
Quello che non sono riuscito a risolvere è secondario, rispetto al "core" dell'impianto (non considerando l'aspetto estetico della plancia su cui quando avrò tempo magari ci rimetterò mano - suggerimenti?) è solo relativo a rendere manipolabile l'`offset` rispetto all'evento `sunrise`.
Ho provato in tutti i modi ma non ce l'ho fatta.
Credo che dovrò cambiare il `trigger` copiando su un sensore il valore dell'alba, manipolarlo per aggiungere/togliere l'offset descritto in un `input_number` e poi confrontarlo con `sensor.time` ma per fare le somme/sottrazioni con le ore ... devo ancora studiare. 
Su questo punto, se ci sono idee più furbe, accetto molto volentieri consigli o riferimenti a documantazione

in ogni caso se il package può interessare lo posso allegare.
ciao

**Reply by serbat** [reply]
Ciao Riccionese
Sto anche io cercando di fare qualcosa di simile al tuo progetto, se posso permettermi un suggerimento, invece che creare 2 tabelle distinte forse è meglio avere una ripetizione di ciclo, ad esempio 15Min totali con 3 ripetizioni da 5Min ciascuno.
In pratica su suggerimento del mio giardiniere ha detto che la soluzione migliore e quella di dare tempi corti per dare tempo all'acqua di impregnare la superficie, ripetendo 3,4 volte il processo secondo lui è il risultato migliore, e questo è quello che sto cercando di fare io, ma sono mooolto niubbo in merito, ma mi ci sono cimentato...
Comunque il package interessa pure a me, tornerebbe utile come esempio visto che la base del progetto da cui sono partito io è la medesima..

**Reply by riccionese** [reply]
Ho fatto qualche ulteriore modifica.
sono riuscito ad unificare gli script per l'avvio su base oraria con quello in base all'alba, cu sono finalmente riuscito a configurare uno slider per gestire l'offset rispetto l'alba stessa. Più altri miglioramenti marginali ora la plancia si vede così:
![image](https://user-images.githubusercontent.com/62329292/130069078-1deb723e-f751-42fb-8a29-b8072dcb42ed.png)

sui consigli dei giardinieri ... sorvolo. :-) ne conosco diversi e non c'è verso che si mettano d'accordo. in generale concordano che non va irrigato in ore troppo calde per vari motivi ma sul resto ognuno ha la sua  !

Nella mia implementazione mi sono basato anche sul funzionamento della centralina che avevo prima e che con questo progetto ho sostituito ma modificare in modo che ci siano più cicli/giorno non è complicato.

per il sorgente del pkg e se vuoi anche della plancia lovelace, se interessa fa sapere  come farle avere

**Reply by essegi65** [reply]
Molto ben fatto.
Interesserebbe anche a me, come grande punto di partenza per adeguarlo al mio impianto, grazie.

**Reply by riccionese** [reply]
Oi. ... eccoli.
come vedrete non è roba complicata ma a me è servito anche per fare un po' di sperimentazione, da niubbo.
C'è qualche cposa che sicuramente si poteva fare meglio ma intanto, per adesso, sono arrivato qui

[irrigazione.yaml.txt](https://github.com/indomus/forum/files/7015503/irrigazione.yaml.txt)
[Plancia_irrigazione.yaml.txt](https://github.com/indomus/forum/files/7015515/Plancia_irrigazione.yaml.txt)

Buon divertimento!
e se vi vengono idee per migliorare ... passatele !
:-)

**Reply by essegi65** [reply]
Stavo guardando il tuo package non ho capito il criterio di scelta tra i giorni della settimana (lu-ma-me-gi-ve-sa-do) e lo slide dell'intervallo dei giorni.
Sicuramente sbaglio ma non sono la stessa cosa (scelta dei giorni in cui far partire l'irrigazione)?
Per il resto posso solo confermarti che mi sembra un ottimo package. Cercherò di adattarlo al mio impianto e se trovo soluzioni utili le condivido

**Reply by riccionese** [reply]
> Stavo guardando il tuo package non ho capito il criterio di scelta tra i giorni della settimana (lu-ma-me-gi-ve-sa-do) e lo slide 
> dell'intervallo dei giorni.
> Sicuramente sbaglio ma non sono la stessa cosa (scelta dei giorni in cui far partire l'irrigazione)?

Circa.
potrebbero essere condizioni sovrapponibili ma le ho tenute distinte per maggior controllo.
Intanto perché i giorni della settimana non consentirebbero una periodicità "esatta" (7 è un numero "scomodo" da dividere in qualunque frazione) ma più che altro i singoli giorni mi consentono un controllo esatto su quali giorni NON avere irrigazione a prescindere da tutte le altre condizioni impostabili.
Se la Domenica non volessi mai irrigare la spengo. 
Così anche se il "ogni X gg" dello slider potrebbe abilitare, il controllo esatto sui giorni lo spegnerebbe.
(anche questa funzione l'ho copiata dalla mia precedente centralina Hunter PSC-Z ed i suoi no-water-days)

in ogni caso tieni anche conto che questo è il risultato "di adesso" su cui ho implementato a cipolla man mano che imparavo e che mi venivano idee ed è molto diverso dalla prima release.
Vedremo se avrò idee e tempo per migliorare ancora.

**Comment by serbat** [comment]
Mitico... riccionese , grazie infinite
Mi sono ritrovato i 2 file nalla mail, grazie ancora
Se non è chiedere troppo, e lo ritieni cosa utile mettere almeno su una scheda "Primo Ciclo" una impup_number per eseguire Tot volte le impostazioni della prima scheda, che so sino a un massimo di 5, cosi è possibile ripetere lo stesso ciclo 5 volte, prima di passare alla seconda scheda....

**Reply by riccionese** [reply]
ieri era il mio ultimo giorno di ferie e quindi anche il tempo di giocherellare con questo pkg calerà verticalmente ma ... ci provo.
non dovrebbe essere difficile aggiungere un `input_number` ed un ciclo

**Reply by riccionese** [reply]
Ho aggiunto un `input_number` per definire le ripetizioni giornaliere.
Non ho previsto un ulteriore campo/slider per definire l'intervallo tra un ciclo ed il successivo ma ho fissato un `delay` di 30 minuti fisso.
se vorrai rendere parametrico anche quello non ci metterai molto.
Adesso la plancia sarebbe così

![image](https://user-images.githubusercontent.com/62329292/130317848-81b7e91a-c1f5-4e6b-9ab8-33cf1eb3e48f.png)

e questi sono i file
[irrigazione_plancia.yaml.txt](https://github.com/indomus/forum/files/7025265/irrigazione_plancia.yaml.txt)
[irrigazione_pkg.yaml.txt](https://github.com/indomus/forum/files/7025266/irrigazione_pkg.yaml.txt)

ciao

**Reply by riccionese** [reply]
e forse più che "Primo Ciclo" e "Secondo Ciclo" ... (ora che ho aggiunto anche le ripetizioni giornaliere) sarebbe più corretto e meno ambiguo descrivere come "Programma A" e "Programma B"

**Comment by serbat** [comment]
Ciao Riccionese
Grazie infinite per l'implementazione, ma non è per caso che per far funzionare il tutto sewrvono anche gli script?

Praticamente gli switc si accendono per una frazione di secondo poi si spengono subito..

**Reply by riccionese** [reply]
non capisco a cosa tu ti riferisca.
gli script sono tutti già dentro il package e gli switch ... quelli nel pkg sono "i miei" che descrivono i relè (sonoff CH4ProR3) che comandano le elettrovalvole in MQTT.

**Reply by serbat** [reply]
Ciao...
sicuramente sbaglio qualcosa, ma ovviamente non capisco quale sia il problema, anche se klicco sulla attivazione manuale in pratica lo switch si abilita per una frazione di secondo mentre in aviio temporizzato impostando l'ora non parte propio.... 
Inoltre non mi trova gli ultimi due input number che hai aggiunto

![immagine](https://user-images.githubusercontent.com/83161647/130323122-5ba83b1c-ccc5-46cb-ab6a-75d94202e8bb.png)


Gli id degli switch li ho sostituiti co uno dei miei...

**Reply by serbat** [reply]
allego immagine della sostituzione switch

![immagine](https://user-images.githubusercontent.com/83161647/130323207-b0f07249-d581-4b00-aab0-f97bb22a3c14.png)

**Reply by serbat** [reply]
... scusa 
trovato il problema sullo spegnimento immediato... siccome questo switch lo sto usando come prova era abilitato anche su un'altra funzione con il tempo impostato a zero, era questo che lo fermava immediatamente...ùscusa ancora sono molto "niubbo" in tutto cio
Ora funziona regolarmente...
manca solo l'imput number

**Reply by serbat** [reply]
trovato, dovevo aggiungere l'integrazione "number-box",,, ora funziona
![immagine](https://user-images.githubusercontent.com/83161647/130323653-12da770b-3779-4d79-b6b4-8f861240ac1d.png)

Scusa se ti tedio in continuazione..... grazie ancora

**Comment by serbat** [comment]
Ri-ciao
Ho risolto il problema dei "input number" mancanti, .... dovevo installare il frontend "Number Box "
![immagine](https://user-images.githubusercontent.com/83161647/130323996-ebae4332-a07e-45c7-9450-6a161b95228d.png)


Come ho risolto il problema del distacco immediato dello switch, in pratica lo switch che uso, lo uso come prova un po in tutto e lo avevo abbinato ad un timer impostato con tempo zero, per quello si spegneva subito...

Continuo con le prove.

**Comment by serbat** [comment]
Ciao Riccionese...
Solo tu mi puoi aiutare.... è tutt'oggi che provo a far funzionare la partenza del ciclo tramite "Orario di irrigazione 1" e "2" ma propio non riesco mi sfugge qualcosa, il bello è che con "attivazione manuale" il ciclo  funziona, ( ho modificato solo i primi 2 switch perchè al momento ne ho solo 2 )

![immagine](https://user-images.githubusercontent.com/83161647/130364676-0e8df892-4591-401a-bcbc-72fc3c4a5c44.png)

Di seguito ti allego anche il file di configurazione...
[irrigazione_pkg-in-uso.yaml.txt](https://github.com/indomus/forum/files/7027836/irrigazione_pkg-in-uso.yaml.txt)

Ancora una volta grazie mille....

**Reply by snakuzzo** [reply]
Prima di tutto...hai controllato la timezone impostata su HA?

**Reply by serbat** [reply]
Ciao snakuzzo
Sinceramente no, e non saprei neppure come controllarla e dove....
Gentilmente, riesci a dirmi cosa controllare....
Grazie

**Reply by snakuzzo** [reply]
Configuration > general

**Reply by serbat** [reply]
questo.... devo posizionarmi sulla mia citta?

![immagine](https://user-images.githubusercontent.com/83161647/130367960-19c84d4b-63fe-4d02-916a-df185a68edbc.png)

**Reply by snakuzzo** [reply]
Quello ok, ma io parlavo della timezone, quella che vedi scritta con FUSO ORARIO.
Devi mettere Europe/Rome

**Reply by Franco2255** [reply]
Ciao
 io ho trovato questo che è molto semplice
![irr](https://user-images.githubusercontent.com/59472878/130672717-26dee391-1399-4e0a-be5d-383c3ab5824a.PNG)

**Reply by Maspar79** [reply]
Scusa Franco ma questo non è mica un package? E' il componente aggiuntivo "Simple Scheduler".

**Reply by Franco2255** [reply]
lo so ma è molto semplice e funzionale

**Comment by serbat** [comment]
Si, ho impostato i valori corretti nel file "configuration.yaml" l'ora adesso è corretta ma l'automazione non parte...

![immagine](https://user-images.githubusercontent.com/83161647/130369131-48e874d6-0932-4cff-bdf6-78e3ac4a9dcf.png)

**Reply by riccionese** [reply]
L'automazione parte in base al `sensor.time` 
prova ad andare in "strumenti per sviluppatori" --> modelli e mettere 
`{{ states('sensor.time') }}` 
e
`{{ states('input_datetime.ora_inizio_ciclo_1')[0:5] }}`
per vedere cosa ti restituisce.se hai lasciato i miei trigger, uso il confronto tra questi due (oltre all'`input_select`)

**Reply by riccionese** [reply]
**ATTENZIONE!!**
nella tua Plancia (che è rimasta uguale alla mia) vedo che hai il "rain-click" con Valore "Piovuto" e questa condizione NON FA PARTIRE le irrigazioni
per cui, se hai una entità testabile che può fungere da sensore pioggia, configurala e testa il valore opportuno,
altrimenti devi togliere questo test dalle varie  ` conditions`, cioè queste righe
```
  - condition: state
    entity_id: sensor.rainclick
    state: 'Asciutto'
```

**Reply by snakuzzo** [reply]
Scusa @serbat, potresti evitare di creare nuovi thread ad ogni risposta? Rispondi sotto l'ultimo messaggio per favore...altrimenti non si capisce niente.

Grazie

**Reply by serbat** [reply]
@snakuzzo  ok, scusa ( se commetto altri sbagli dimmelo pure tranquillamente, altrimenti non imparo nulla )

@riccionese 
Grazie mille,infatti ho commentato "#" tutte le 3 righe dove si presentavano, ora parte finisce il primo ciclo, ma non esegue le ripetizioni giorno.

Una cosa secondo me utile da implementare potrebbe essere quella di far comparire il tempo di fine dei cicli comprese le ripetizioni giorno in modo da avere sott'occhio quando realmente completa tutti i cicli... magari mettendolo in questa zona senza impegnare ulteriore posto

![immagine](https://user-images.githubusercontent.com/83161647/130435945-fb56d8b2-c2aa-4199-88e8-709804e29882.png)

**Reply by riccionese** [reply]
Mi fa piacere che tu abbia risolto almeno lo "start" 
Per le ripetizioni non credo di poterti aiutare molto. Lo script che è nel pkg io l'ho testato (anche se non mio impianto non lo uso) e "ripete".
In ogni caso, davvero senza nessun intendo polemico, se vai avanti con il copia&incolla non imparerai molto.
Nei file che ho allegato le condizioni in base alle quali "succedono cose" o gli script o le automazioni ... sono in chiaro. 
Non sono bellissime, (probabilmente se @snakuzzo ci guardasse dentro mi picchierebbe sulle nocche delle dita :-) ) e ho già idea di qualche miglioramento, soprattutto sulla riduzione della sezione che testa singolarmente ogni giorno della settimana. 
Vorrei provare a fare uno script generico a cui passare come variabili i giorno e gli altri parametri delle condizioni e per me questa è "palestra e stimolo" per capire e realizzare... che è quello che suggerirei anche a te.
Comunque accetto lo stimolo del tuo suggerimento di aggiungere un timer per il ciclo, e ci proverò ma non garantisco :-)

**Reply by serbat** [reply]
@riccionese 
Non posso che darti ragione,provo... vediamo che succede..
Grazie ancora.

**Reply by riccionese** [reply]
Ho provato a realizzare un timer con durata definita da variabili
questo il pgk di test che ho appena provato e mi funziona
```
timer:
  timer_test:
    name: test timer

input_number:
  timer_durata:
    name: Durata Timer
    initial: 0
    min: 0
    max: 100
    step: 1
    mode: box

automation:
  - id: timer_test_on
    alias: accendo il timer
    trigger:
      - platform: template
        value_template: "{{ states('input_number.timer_durata') | int > 0 }}"
    action:
# Questo servizio lo fa partire usando il timer_durata come parametro
# invece del vlore input_number di questo esempio, puoi usare i valori dgli input_number delle durate dei vari irrigatori.
# occhio che le durate degli irrigatori sono in minuti menter qui sono secondi.
      - service: timer.start
        data_template:
          entity_id: timer.timer_test
          duration: "{{ states('input_number.timer_durata') }}"
# Questo riazzera il timer_durata, visto che come esempio di trigger ho usato il cambio di valore da zero a maggiore di zero
# nel caso reale non dovrebbe servirti, ché altrimenti azzeri le programmazioni
      - service: input_number.set_value
        entity_id: input_number.timer_durata
        data:
          value: 0
```

pistolandoci un po' dovresti poterlo inserire nel tuo pkg in modo da poter aggiungere nella plancia la `entity: timer.timer_test` (ovviamente con il nome del/i tuo/tuoi timer)

Io non lo  messo nel mio. non mi serve (non guardo la plancia mentre irriga :-) ) 
vedi tu se ti può essere di aiuto
ciao

**Reply by riccionese** [reply]
alla fine ci ho voluto provare lo stesso anche io.

![image](https://user-images.githubusercontent.com/62329292/130486055-b522154d-88ac-499f-b4b5-69856a9289df.png)
![image](https://user-images.githubusercontent.com/62329292/130486120-04fc03e8-63ee-430a-a502-dff623574f51.png)

Nota: 
1. non conteggio i 5 secondi di delay tra una zona e l'altra
2. conteggio la durata di un solo ciclo della programmazione. Se hai le ripetizioni e vuoi il tempo totale ... devi fare le opportune modifiche al calcolo

[irrigazione_pkg.yaml.txt](https://github.com/indomus/forum/files/7033659/irrigazione_pkg.yaml.txt)
[irrigazione_plancia.yaml.txt](https://github.com/indomus/forum/files/7033660/irrigazione_plancia.yaml.txt)

buon divertimento
:-)

Ciao

**Reply by serbat** [reply]
Grande riccionese...
mi era cimentato pure io per farlo, ma con esito negativo, mille grazie comunque per l'implementazione.

Ho notato che le ripetizioni ( sul precedente ) sono solo 2, ho trovato come aggiugerne altre ma al momento slegate dal slider, cercherò di implementare la funzione "until".

Confesso che pur avendo una certa dimestichezza con VB6, con questo linguaggio ancora non sono riuscito a capirne a fondo la sequenza logica, e mi manca tantissimo un debugger per controllare il codice riga per riga...... che tu sappia esiste un debug per visionare il codice con la possibilità di verifica riga per riga.... con questo sistema pur dandomi ok al codice di fatto poi non funziona e non riesco a risalire all'errore...il tutto è un pò frustante almeno finche non arriverò a capire la logica, forse sono abituato male...

**Reply by riccionese** [reply]
Non conosco debugger. 
ma in "strumenti per sviluppatori" --> "modelli" puoi già fare molto. 
Anche io sono abbastanza agli inizi e sperimentando in "modelli" e leggendo le varie documentazioni (nella stessa pagina "modelli" ci sono 2 link utilissimi). 
Poi "rubando" dai lavori di altri e cercando di indagarli sperando di capirli (non sempre) e provando a reimplementare quello che "credo" di aver capito ... passettin passettino mi sto muovendo.
Questo stesso progetto per l'irrigazione l'ho iniziato partendo da un pkg trovato su un forum.
Ora il mio è sostanzialmente diverso (e penso migliore, ora) ma senza il bootstrap avuto con quella fonte non avrei saputo come iniziare
Una mia esperienza: trovo molto utile la gestione dei pkg. Ho sempre un "prova.yaml" dove faccio test spiccioli ma quando voglio salire di complessità ne creo uno nuovo dedicato, come ho fatto con il discorso del timer (... per il quale, a sua volta, mi sono "ispirato" al pkg per lo zigbee2mqtt preso dalla guida su inDomus.it :-) )
In questo modo i pasticci che posso fare sono sicuramente solo lì dentro e basta cambiare l'estensione e riavviare per includerlo/escluderlo.

... e poi c'è sempre il forum e telegram per chiedere consigli e supporto !

buon divertimento !

**Reply by serbat** [reply]
Ho provato il nuovo script con il tempo totale, funziona, ma intendevo una cosa diversa, cioè avere il tempo totale di quando finisce 
es: inizio a irrigare alle 20:00 sommando tutte le irrigazioni mi dirà che finiro per le ore 24:00, non il conto alla rovescia

Questo torna utile per avere ad occhio il tempo totale di irrigazione, sopra tutto in quei comuni ( come il mio ) che a determinati orari è vietato innaffiare e se dopo le ore 6 ho il divieto, questa funzione mi aiuta a non "sforare" involontariamente il tempo, ........ questo vorrei
provarlo a risolvere io :)

**Reply by riccionese** [reply]
ah. capito.
comunque non ti dovrebbe essere difficile: gli elementi li hai tutti, va fatta solo una somma ed una conversione da secondi/timestamp a data-ora
o poco più
good luck!

**Reply by Maspar79** [reply]
Complimenti riccionese per il tuo package, ricordo la discussione del mese scorso in cui eri alla ricerca di un dispositivo con quattro switch ... adesso hai appena realizzato tutto. Proverò quanto prima il tuo lavoro, ti farò sapere.

**Reply by serbat** [reply]
Finalmente...... anche ase a fatica sono riuscito a fare qualcosa.

![immagine](https://user-images.githubusercontent.com/83161647/130848593-72754b78-ccd6-4fe7-9617-e3834ec9c667.png)


Pero non sono riuscito a ridurre la dimensione ( spazio ) della input_text evidenziata.... c'è un sistema?

Ora provo a vedere se riesco a fare un ciclo per le ripetizioni di irrigazione giorno utilizzando un ciclo tipo "until" o "while" che esegua le ripetizioni pianificate...

Scusate..... ho rifatto il messaggio, avevo sbagliato nella replica.... sorry

**Comment by Semex82** [comment]
Ciao a tutti, ho scoperto per caso questo sito perché cercavo un modo per fare l'irrigazione del mio giardino con la particolarità che dovrei usare una doppia fonte di alimentazione (Depuratore e qualora non bastasse l'acqua integrare da rete) e questa base è fantastica penso partirò da questa. Prima di imbarcarmi però chiedo se a vostro avviso posso implementarla con:

- sensore di livello serbatoio depuratore (Se non è sufficiente attiva l'elettrovalvola per usare acqua di rete)
- sensore umidità per irrigare in base all'umidità del terreno
- Usare 2 CH4 in modo da poter dividere in almeno 6 zone
  
Ovviamente condividerò gli eventuali sviluppi con tutti voi

Grazie

**Comment by effe-7** [comment]
ciao a tutti, sarebbe possibile avere il file del progetto ?

grazie

**Reply by indomus** [reply]
https://github.com/indomus/forum/discussions/584#discussioncomment-1207051

**Reply by marco-dea** [reply]
Ciao.
Mi potresti far sapere che valvole hai utilizzato
Grazie e complimenti per il progetto
